### ISSUE TITLE
Decide what to do about undocumented `Witness::push` panics

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#3302]
These panics are caused by us storing indices as `u32` rather than `usize` and we forgot to document it.

Possible solutions:

* store `usize` indices
* keep panics and document them, also make them more explicit with `track_caller` etc.
* make `push` fallible
* require `PushBytes` instead of `[u8]`

See https://github.com/rust-bitcoin/rust-bitcoin/pull/3259#issuecomment-2329605961

**Comment by apoelstra** [comment]
Storing `usize` indices just pushes the problem further out -- if you go out of range you'll then have `Witness`es that nodes will not parse. (Of course, if you have 4GB+ witnesses you'll likely have many other problems so maybe this is an academic concern..)

I would suggest making `push` fallible, and providing another method to push objects that are less than MAX_SIZE large. (I'm not sure if `PushBytes` is the right type? I feel that `Script` should be, and we should also have a MAX_SIZE limit on `Script`...though that would undermine our ability to convert to/from `Vec`s...)

**Comment by apoelstra** [comment]
>Is the API complexity somehow worth it here?

Yes, because this API is less-used and the required logic is way simpler (you literally just check the length on each push and reject the push if it's over). It's the same complexity as with `PushBytes` except arguably less disruptive.

With `Transaction` you have multiple fields per input and per output and can't even compute the total length without reserializing. You'd basically need an opaque type, a `Transaction::try_update` method which took a closure that was passed a copy of the transaction to modify, and after the closure finished the copy was reserialized and accepted/rejected.

>Maybe the type should be renamed to Max32bLenBytes (or a better name) and used for both.

Yeah, we would want a better name (and the actual limit, if we are taking compact-size decodability seriously is 32 *megs*). TBH we could just keep the name `PushBytes` and reduce its size limit. It looks like `script::Builder::push` already takes an `AsRef<PushBytes>` so there'd be no change there; we'd just start using the same bound for `Witness::push`.

As an aside, in the docs for these functions we should mention that the docs for `PushBytes` have a whole list of `AsRef<PushBytes>` implementors. (I didn't know this until just now when I checked.)

> That means all push operations [on script::Builder] become fallible...

Yeah, I see. For scripts you're right. With `Witness` we are only checking the length of individua pushes, meaning that we can have infallible `push_signature`, `push_pushbytes`, etc., methods. But if users had to unwrap every `.push_opcode` on Script that would be extremely annoying (and useless, because as you say, everyone would just unwrap).

**Comment by apoelstra** [comment]
> The issue with `Witness` is the same as `txins` - when pushing you may still exceed the total limit on transaction

This is true, but there's still value in checking that the `Witness` itself is in limit. (Specifically, the value is that we don't need to think about how to handle out-of-range compact sizes *within the `Witness` type itself*.)

> If the script is required be within some size then it makes sense to restrict `PushBytes` to that size instead of 2^32B. Re-adds branches on 32-bit architectures but it's fine since we never promised there won't be any.

Yep -- scripts are always serialized with a compact-size prefix, so they are subject to the 32M limit.

> There's still the possibility that one pushes too many items into the `Witness`. I'd say use `PushBytes` as a best-effort protection and give up on other things.

Agreed. I am a little tempted to add a `cfg` flag that tries to panic when it detects that the total size of `Witness`es or `Script`s exceeds the limit, which may help some people during debugging. But for normal development I think we would need dependent types to do this "ergonomically".

**Comment by Kixunil** [comment]
> Specifically, the value is that we don't need to think about how to handle out-of-range compact sizes _within the `Witness` type itself_.

We could just use `usize` but surely it'd waste a bit of space if such witnesses are invalid to begin with. So I'd argue the real value of requiring `PushBytes` is a memory/cache optimization. :)

> I am a little tempted to add a `cfg` flag that tries to panic when it detects that the total size of `Witness`es or `Script`s exceeds the limit

Why not just `debug_assert`? I think this is quite in-line with with Rust does with integers.

> we would need dependent types to do this "ergonomically".

I suspect even that wouldn't really help that much.

Anyway, I plan for `Witness` to have a bunch of standard constructors so unless you're doing custom scripts you don't have to deal with any of this. (And even then I want appending the control block & co to be easier.)

**Comment by tcharding** [comment]
> (and the actual limit, if we are taking compact-size decodability seriously is 32 megs)

When you say 32 megs, are you referring to the ` uint64_t MAX_SIZE = 0x02000000` or some other limit?

**Comment by apoelstra** [comment]
> When you say 32 megs, are you referring to the ` uint64_t MAX_SIZE = 0x02000000` or some other limit?

Yes, that is what I'm referring to.

> Why not just `debug_assert`? I think this is quite in-line with with Rust does with integers.

I worry that this could still turn an "object that can't be validly serialized' bug into a DoS bug. But yeah, I wouldn't be opposed to doing this.

**Comment by Kixunil** [comment]
Well, `debug_assert` is configurable via a canonical cfg and most people don't use debug asserts in production.

**Comment by apoelstra** [comment]
Hmm, ok, yeah, I'm warming up to the idea to just using debug_assert.
